File: java/src/main/java/org/revenj/patterns/DomainEventStore.java
package org.revenj.patterns;

import java.util.Collections;
import java.util.List;

public interface DomainEventStore<TEvent /*extends DomainEvent*/> extends Repository<TEvent> {
	String[] submit(List<TEvent> domainEvents);

	default String submit(TEvent domainEvent) {
		return submit(Collections.singletonList(domainEvent))[0];
	}

	void mark(String[] uris);

	default void mark(String uri) {
		mark(new String[]{uri});
	}
}

File: scala/revenj-core/src/main/scala/net/revenj/patterns/DomainEventStore.scala
package net.revenj.patterns

import scala.concurrent.Future

trait DomainEventStore[T <: DomainEvent] extends Repository[T] with SearchableRepository[T] {

  def submit(events: Seq[T]): Future[IndexedSeq[String]]

  def submit(event: T): Future[String] = {
    import scala.concurrent.ExecutionContext.Implicits.global
    require(event ne null, "null value provided for event")
    submit(Seq(event)).map(_.head)
  }

  def mark(uris: Seq[String]): Future[Unit]

  def mark(uri: String): Future[Unit] = {
    require(uri ne null, "null value provided for URI")
    require(uri.length != 0, "empty value provided for URI")
    mark(Seq(uri))
  }
}
